docstring-formatter.lisp

```diff
@@ -0,0 +1,133 @@
+            (form list))
+    (let ((order (visiting-order generator node))
+          (before nil))
+      (dolist (next order)
+        (let ((data (getf form next)))
+          (visit-one output
+                     data
+                     generator
+                     next
+                     node
+                     before)
+          (unless (null data)
+            (push next before)))))
+    output))
+
+(defclass basic-docstring-formatter (fundamental-docstring-formatter)
+  ())
+
+
+(let ((node (make-instance 'fundamental-type-node)))
+  (defmethod type-node-instance ((generator basic-docstring-formatter)
+                                 (type symbol))
+    node))
+
+
+(defmacro define-stream-visitors
+  ((output output-class) (generator generator-class)
+   node label data before
+   &body body)
+  `(progn
+     ,@(mapcar (lambda (x)
+                 (destructuring-bind ((node-class label-symbol d-class) . body) x
+                   `(defmethod visit-one ((,output ,output-class)
+                                          (,data ,d-class)
+                                          (,generator ,generator-class)
+                                          (,label (eql ,label-symbol))
+                                          (,node ,node-class)
+                                          (,before list))
+                      (unless (endp ,before)
+                        (format output "~%~%"))
+                      ,@body)))
+               body)))
+
+
+(define-stream-visitors
+    (output stream)
+    (generator basic-docstring-formatter)
+    node label data before
+  ((fundamental-type-node :side-effects string)
+   (format output "Side Effects:~% ~a" data))
+  ((fundamental-type-node :side-effects list)
+   (format output "Side Effects:~%~{ * ~a~%~}" data))
+  ((fundamental-type-node :arguments list)
+   (format output "Arguments:~%")
+   (loop for d on data
+         do (destructuring-bind (arg desc) (first d)
+              (format output " ~a: ~a" arg desc)
+              (unless (endp (rest d))
+                (format output "~%")))))
+  ((fundamental-type-node :returns string)
+   (format output "Returns:~% ~a" data))
+  ((fundamental-type-node :returns list)
+   (format output "Returns:~%~{ * ~a~^~%~}" data))
+  ((fundamental-type-node :notes string)
+   (format output "Note:~% ~a" data))
+  ((fundamental-type-node :notes list)
+   (format output "Notes:~%~{ * ~a~^~%~}" data))
+  ((fundamental-type-node :description string)
+   (format output "~a" data))
+  ((fundamental-type-node :examples list)
+   (format output "Examples:~%~{ ~a~^~%~^~%~}" data))
+  ((fundamental-type-node :examples string)
+   (format output "Example:~% ~a" data))
+  ((fundamental-type-node :exceptional-situations list)
+   (format output "Exceptional Situations:~%~{ * ~a~^~%~}" data))
+  ((fundamental-type-node :exceptional-situations string)
+   (format output "Exceptional Situations:~% ~a" data)))
```

Comment:
I honestly can't quite tell what all this visitor stuff is supposed to be for. What was your thought process behind this? What does this design accomplish aside from complicating the design in favour of a very particular kind of formatting syntax?

Reply:
Idea is to keep formatter both extendable and customizable. This way user can change style that is supposed to be used in docstrings by the means of formatter protocol. Formatter is here just to have simple way to dispatch logic.